### PR TITLE
Addtime, Subtime functions

### PR DESCRIPTION
**Pull request by Tratar**
Implemented ADDTIME, SUBTIME see here: https://github.com/pingcap/tidb/issues/236

### CODE REVIEW

**Comment by Tratar** [comment]
Get rid of fmt at all. Conflict resolved

File: expression/builtin.go [hunk 1]
```diff
@@ -174,6 +173,8 @@ var funcs = map[string]functionClass{
 	ast.Curtime:          &currentTimeFunctionClass{baseFunctionClass{ast.Curtime, 0, 1}},
 	ast.Date:             &dateFunctionClass{baseFunctionClass{ast.Date, 1, 1}},
 	ast.DateAdd:          &dateArithFunctionClass{baseFunctionClass{ast.DateAdd, 3, 3}, ast.DateArithAdd},
+	ast.AddTime:          &timeArithFunctionClass{baseFunctionClass{ast.AddTime, 2, 2}, ast.TimeArithAdd},
```
Comment (tiancaiamao) [inline expression/builtin.go]: the old line is sorted by function name...

**Comment by tiancaiamao** [comment]
Time function is hard due to some compatibility reason.

As you will see, MySQL behavior is strange:

```
mysql> SELECT ADDTIME('2007-12-00 00:00:00', '1 1:1:1.000002');
+--------------------------------------------------+
| ADDTIME('2007-12-00 00:00:00', '1 1:1:1.000002') |
+--------------------------------------------------+
| 2007-12-01 01:01:01.000002                       |
+--------------------------------------------------+
1 row in set (0.00 sec)
```

I suggest you can have a try with just `Subtime` function, and call 
`Time.Sub` in "github.com/pingcap/tidb/util/types" package to implement it.

**Comment by Tratar** [comment]
> As you will see, MySQL behavior is strange:
Why the behaviour is strange? I don't actually understand.

**Comment by tiancaiamao** [comment]
Go time library support neither this time value '2007-12-00', nor add operation on it.
December doesn't have zero day, but this date is valid in MySQL.

**Comment by Tratar** [comment]
I see, but my implementation shows the same result. I added two tests with illegal dates. It works like mysql does.
{"2007-12-00 00:00:00", "1:1:1.000002", "2007-11-30 01:01:01.000002", "2007-11-29 22:58:58.999998"},
{"2007-12-00 00:00:00", "1 1:1:1.000002", "2007-12-01 01:01:01.000002", "2007-11-28 22:58:58.999998"},

**Comment by Tratar** [comment]
I'd like to explay, first we parse date like "2007-12-00 00:00:00" to mysqlTime where day == 0 is ok, then it adds/subs duration and creates types.Time instance with correct date(like "2007-11-30 00:00:00"), then it converts date back to mysqlTime and wraps it to types.Time(like datum like).

File: expression/builtin_time_test.go [hunk 2]
```diff
@@ -1014,6 +1014,65 @@ func (s *testEvaluatorSuite) TestTimestamp(c *C) {
...
+		t1        string
+		t2        string
+		expectAdd string
+		expectSub string
+	}{
+		{"2007-12-31 23:59:59.999999", "23:59:59.999999", "2008-01-01 23:59:59.999998", "2007-12-31 00:00:00"},
+		{"9207-12-31 23:59:59.1", "-120:44:33", "9207-12-26 23:15:26.100000", "9208-01-06 00:44:32.100000"},
+		{"9207-12-31 23:59:59", "120:44:33.1", "9208-01-06 00:44:32.100000", "9207-12-26 23:15:25.900000"},
+		{"2007-12-00 00:00:00", "1:1:1.000002", "2007-11-30 01:01:01.000002", "2007-11-29 22:58:58.999998"},
+		{"2007-12-00 00:00:00", "1 1:1:1.000002", "2007-12-01 02:01:01.000002", "2007-11-28 22:58:58.999998"},
```
Comment (tiancaiamao) [inline expression/builtin_time_test.go]: fix CI
s/2007-12-01 02:01:01.000002/2007-12-01 01:01:01.000002

File: util/types/time.go [hunk 3]
```diff
@@ -823,6 +823,11 @@ func (d Duration) MicroSecond() int {
 	return frac
 }
 
+//ParseDuration(test.Input, MinFsp)
+// ParseDuration parses the time form a formatted string with a fractional seconds part,
+// returns the duration type Time value.
+// See http://dev.mysql.com/doc/refman/5.7/en/fractional-seconds.html
+
 // ParseDuration parses the time form a formatted string with a fractional seconds part,
```
Comment (tiancaiamao) [inline util/types/time.go]: repeated?

File: expression/builtin_time.go [hunk 4]
```diff
@@ -1479,6 +1479,110 @@ func (b *builtinDateArithSig) eval(row []types.Datum) (d types.Datum, err error)
...
+func timeArithFuncFactory(op ast.TimeArithType) BuiltinFunc {
+	return func(args []types.Datum, ctx context.Context) (d types.Datum, err error) {
+		// args[0] -> DateTime, Time.
+		// args[1] -> Time.
+		// result type equals to args[0] type.
+		if args[0].IsNull() || args[1].IsNull() {
+			return d, nil
+		}
+
+		//todo fsp
```
Comment (tiancaiamao) [inline expression/builtin_time.go]: // TODO: consider fsp

all todo should be uppercase.

**Comment by ngaut** [comment]
Please resolve conflicts.

**Comment by coocood** [comment]
@Tratar 
I'm closing this for now since it has been too long without any updates.

**Pull request closed without merging**